cpp/src/round/round.cu

```diff
@@ -47,102 +52,190 @@ T __device__ generic_abs(T value)
 }
 
 template <typename T, typename std::enable_if_t<std::is_signed<T>::value>* = nullptr>
-bool __device__ is_negative(T value)
+int16_t __device__ generate_sign(T value)
 {
-  return value < 0;
+  return value > 0 ? 1 : value < 0 ? -1 : 0;
 }
 
 // this is needed to suppress warning: pointless comparison of unsigned integer with zero
 template <typename T, typename std::enable_if_t<not std::is_signed<T>::value>* = nullptr>
-bool __device__ is_negative(T value)
+int16_t __device__ generate_sign(T value)
 {
-  return false;
+  return value > 0 ? 1 : 0;
 }
 
-struct round_fn {
-  template <typename T, typename... Args>
-  std::enable_if_t<not cudf::is_numeric<T>(), std::unique_ptr<column>> operator()(Args&&... args)
+template <typename T>
+constexpr inline auto is_supported_round_type()
+{
+  return cudf::is_numeric<T>() && not std::is_same<T, bool>::value;
+}
+
+template <typename T>
+struct half_up_zero {
+  T n;  // unused in the decimal_places = 0 case
+  template <typename U = T, typename std::enable_if_t<cudf::is_floating_point<U>()>* = nullptr>
+  __device__ U operator()(U e)
   {
-    CUDF_FAIL("Type not support for cudf::round");
+    return generic_round(e);
   }
 
-  template <typename T>
-  std::enable_if_t<cudf::is_floating_point<T>(), std::unique_ptr<column>> operator()(
-    column_view const& input,
-    int32_t decimal_places,
-    cudf::rounding_method method,
-    cudaStream_t stream,
-    rmm::mr::device_memory_resource* mr)
+  template <typename U = T, typename std::enable_if_t<std::is_integral<U>::value>* = nullptr>
+  __device__ U operator()(U e)
+  {
+    assert(false);  // Should never get here. Just for compilation
+    return U{};
```

Comment:
Can you successfully compile if you replace these two lines with `CUDF_FAIL(<msg>)`? I think we do this elsewhere. That would make these into one-liners.

Reply:
You can't `CUDF_FAIL` in device code.